### DISCUSSION TITLE
How to handle reserved transactions via Nordigen?

### DISCUSSION THREAD

**Discussion by alekslyse** [firefly-iii/firefly-iii#8094] [General]
As nordigen exposes reserved transactions it gives the importer noe enough information to properly propergate the information in firefly iii. Is it possible to 1. ignore the transactions that is reserved and import them when they are completed or 2. import the reserved, but automatically update them when they become final?

At least for Norwegian bank accounts the transactions can be reserved for anything between hours to days, so its hard to know when the reservation ends (its in the Nordigen API of course).

**Comment by JC5** [comment]
Updating the transactions will not really work. 

Is the transaction actually imported or does it fail?

**Reply by alekslyse** [reply]
Actually the transaction is imported with no issue, but its missing some information like destination account, and some details (the title is usually reserved), so automations etc doesnt work properly). All the info is not propagated to the transaction before it is cleared in the bank. So if it was at least possible to wait with uncleared transactions it would be the best.

All the uncleared is under the "pending" value of the nordigen xml and they all got transactionId of 0, so potential for double import when it moves to cleared. 

My suggestion is that we can ignore the pending transactions, or transactions with transactionId 0

**Reply by alekslyse** [reply]
Here is an example of an pending transaction

`   "pending": [
            {
                "transactionId": "0",
                "entryReference": "1486",
                "bookingDate": "2023-11-02",
                "transactionAmount": {
                    "amount": "-122.56",
                    "currency": "NOK"
                },
                "debtorAccount": {
                    "bban": "xxxxxxxxxxxx"
                },
                "remittanceInformationUnstructured": "CRV*AMAZON PRIME AMZN.CO.",
                "additionalInformation": "ProprietaryBankTransactionText:Varekjøp",
                "proprietaryBankTransactionCode": "946",
                "debtorAgent": "SPAVNOBB"
            },`

**Reply by JC5** [reply]
You can write a rule to delete these transactions. They will not be imported again until they're final.

**Reply by alekslyse** [reply]
How would you suggest to do that? Is it possible to trigger a rule while its importing, so if for example the transactionid = 0 then delete it?

**Reply by JC5** [reply]
Yes it is.

**Comment by poudenes** [comment]
Hi All,

Here in the Netherlands we have also "reservations" for example UberEats. When Firefly imported runs it handle those as normal payments. Then couple of days later the reservation become a payment and get the new date. Then Firefly importer again import this payment. So sometimes I have double payments in my list.

I know that this happen so I check always double when I have a reservation in the list on my mobile banking app.
When I have a new one I will try to see what Nordigen will show me in the transactions information. Maybe we can use some info

**Reply by poudenes** [reply]
Checked logs and found this line

```
Apple Pay",
"payment_date":"2024-01-25",
"order":0,
"currency_code":"EUR",
"tags":["pending"],
"category_name
":null,
"category_id":null,
"notes":"",
"external_id":"010304324980104450000000",
"internal_reference":"",
"additional-information":"",
"source_id":9,
"destination_name":"TooGoodToG 35ggjjp8q7w",
"destination_number":""}  
```

There is a :["pending"] part also in the line. So I can create a rule also for this :). Means remove payment when pending is in the tags field?

<img width="799" alt="Screenshot 2024-01-28 at 10 02 14" src="https://github.com/firefly-iii/firefly-iii/assets/4399532/83d71a28-52bf-475e-9f80-ac36339903c3">

or

<img width="754" alt="Screenshot 2024-01-28 at 10 03 22" src="https://github.com/firefly-iii/firefly-iii/assets/4399532/03f9dfd7-bcdb-4a95-b702-5c051f274254">

**Reply by JC5** [reply]
Yes that should work. The bottom one.

**Reply by poudenes** [reply]
Great. Maybe a "settings" idea. Drop transactions when they are "pending". But first you have to know if every bank use the same way to show this.

**Comment by poudenes** [comment]
Seems not to work. The last line is only a trigger if a transaction has a tags... its not looking to if it have "pending"

**Reply by JC5** [reply]
Then it's "Any tag is..."

**Reply by poudenes** [reply]
Yes.. Didn't saw that one 👍 . Added the rule and lets see what happen!
